src/persist/src/operators/source.rs

```diff
@@ -59,72 +61,131 @@ where
         // it for the operator.
-        let (ok_new, err_new_decode) =
-            operator::source(self, "PersistedSource", |capability, info| {
-                let worker_index = self.index();
-                let activator = self.activator_for(&info.address[..]);
-                let mut cap = Some(capability);
-                move |output| {
-                    let mut done = false;
-                    if let Some(cap) = cap.as_mut() {
-                        let mut session = output.session(cap);
-                        match listen_rx.try_recv() {
-                            Ok(e) => match e {
-                                ListenEvent::Records(mut records) => {
-                                    // TODO: This currently works by only emitting the persisted data on worker
-                                    // 0 because that was the simplest thing to do initially. Instead, we should
-                                    // shard up the responsibility between all the workers.
-                                    if worker_index == 0 {
-                                        session.give_vec(&mut records);
+        let (ok_all, err_all) = match snapshot {
+            Err(err) => {
+                (
+                    operator::empty(self),
+                    // TODO: Figure out how to make these retractable.
+                    vec![(format!("replaying persisted data: {}", err), 0, 1)].to_stream(self),
+                )
+            }
+            Ok(snapshot) => {
+                let snapshot_upper = snapshot.upper();
+
+                // listen to new data
+                let (ok_new, err_new) = listen_source(self, snapshot_upper, listen_rx);
+                let err_new_decode = err_new.flat_map(std::convert::identity);
+
+                // Replay the previously persisted data, if any.
+                let (ok_previous, err_previous) = operators::replay(self, snapshot);
+
+                let ok_all = ok_previous.concat(&ok_new);
+                let err_all = err_previous.concat(&err_new_decode);
+                (ok_all, err_all)
+            }
+        };
+
+        let err_all = err_all.concat(&err_new_register);
+
+        (ok_all, err_all)
+    }
+}
+
+/// Creates a source that listens on the given `listen_rx` and gates (filters out)
+/// updates by the given `gate_frontier`.
+fn listen_source<S, K, V>(
+    scope: &S,
+    lower_filter: Antichain<u64>,
+    listen_rx: Receiver<ListenEvent<Result<K, String>, Result<V, String>>>,
+) -> (
+    Stream<S, ((K, V), u64, isize)>,
+    Stream<S, Vec<(String, u64, isize)>>,
+)
+where
+    S: Scope<Timestamp = u64>,
+    K: TimelyData + Codec + Send,
+    V: TimelyData + Codec + Send,
+{
+    let source_stream = operator::source(scope, "PersistedSource", |mut capability, info| {
+        let worker_index = scope.index();
+        let activator = scope.activator_for(&info.address[..]);
+
+        // For Antichain<u64> there can only ever be one element, because
```

Comment:
Fascinating, I think you're right. My mental model was that Capability was/has an Antichain, but looks like that's wrong. What you have here is good for now